cdap-watchdog/src/main/java/co/cask/cdap/logging/clean/FileMetadataScanner.java

```diff
@@ -0,0 +1,236 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.clean;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetManager;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scan;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.logging.meta.LoggingStoreTableUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import org.apache.tephra.TransactionFailureException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to scan and also delete meta data
+ */
+public class FileMetadataScanner {
+  private static final Logger LOG = LoggerFactory.getLogger(FileMetadataScanner.class);
+  private static final byte[] OLD_ROW_KEY_PREFIX = LoggingStoreTableUtil.OLD_FILE_META_ROW_KEY_PREFIX;
+  private static final byte[] OLD_ROW_KEY_PREFIX_END = Bytes.stopKeyForPrefix(OLD_ROW_KEY_PREFIX);
+
+  private static final byte[] NEW_ROW_KEY_PREFIX = LoggingStoreTableUtil.NEW_FILE_META_ROW_KEY_PREFIX;
+  private static final byte[] NEW_ROW_KEY_PREFIX_END = Bytes.stopKeyForPrefix(NEW_ROW_KEY_PREFIX);
+
+  // cut-off time discount from actual transaction timeout
+  private static final int CUTOFF_DISCOUNT = 5;
+  private final Transactional transactional;
+  private final DatasetManager datasetManager;
+
+  public FileMetadataScanner(DatasetManager datasetManager, Transactional transactional) {
+    this.transactional = transactional;
+    this.datasetManager = datasetManager;
+  }
+
+  /**
+   * scans for meta data in old format which has expired the log retention.
+   */
+  @VisibleForTesting
+  List<byte[]> scanAndDeleteOldMetaData(int transactionTimeout, final int cutoffTransactionTime) {
+    final List<byte[]> deletedEntries = new ArrayList<>();
+    try {
+      transactional.execute(transactionTimeout, new TxRunnable() {
+        public void run(DatasetContext context) throws Exception {
+          Stopwatch stopwatch = new Stopwatch();
+          stopwatch.start();
+          Table table = LoggingStoreTableUtil.getMetadataTable(context, datasetManager);
+          // create range with tillTime as endColumn
+          Scan scan = new Scan(OLD_ROW_KEY_PREFIX, OLD_ROW_KEY_PREFIX_END, null);
+          try (Scanner scanner = table.scan(scan)) {
+            Row row;
+            while ((row = scanner.next()) != null) {
+              if (stopwatch.elapsedTime(TimeUnit.SECONDS) > cutoffTransactionTime) {
```

Comment:
You can move this to the `while` condition. Also, better to have this checked first so that the `scanner.next()` won't get called and the result thrown away. Also, the comparison should be `elapseTime >= maxTime`, otherwise you'd be spending an extra second).

```java
while (stopwatch.elapsedTime(TimeUnit.SECONDS) < maxTxTime && ((row = scanner.next()) != null) {
  ..
}